packages/@aws-cdk/aws-ec2/lib/vpc.ts

```diff
@@ -698,8 +698,22 @@ export class Vpc extends VpcBase {
 
   /**
    * Import an existing VPC from by querying the AWS environment this stack is deployed to.
+   *
+   * Calling this method will lead to a lookup when the CDK CLI is executed.
+   * You can therefore not use any values that will only be available at
+   * CloudFormation execution time (i.e., Tokens).
+   *
+   * If you are looking to share a VPC between stacks, you can pass the `Vpc`
+   * object between stacks and use it as normal.
    */
   public static fromLookup(scope: Construct, id: string, options: VpcLookupOptions): IVpc {
+    if (Token.isUnresolved(options.vpcId)
```

Comment:
Geez. It looks we may be missing this check in tons of places throughout our constructs.

For v2, I wonder if we should strongly type all properties to be `string | IResolvableToken` (hypothetical type - may or may not exist yet) to ensure that constructs deal with both cases?

Reply:
Yes we are. For many places it doesn't matter though (makes it harder!).

This is again an ancient discussion :).

We explicitly opted to do it this way (encode "magic" values into primitive types) in order to guarantee maximum compatibility with all programming languages. Sum types are not supported in many languages, and emulating them would become unwieldy very quickly (especially if you have to have additional syntax for LITERALLY every property you pass).